src/ast.ts

```diff
@@ -399,40 +425,68 @@ export class Symbols {
       isType: <ExpType> isType,
       methods: methods
     };
-    this.register('paths', '/' + parts.join('/'), p);
+    return <Path> this.register('paths', '/' + parts.join('/'), p);
   }
 
-  registerSchema(name: string, derivedFrom: ExpType | ExpUnionType | void, properties, methods) {
-    methods = methods || {};
+  registerSchema(name: string,
+                 derivedFrom?: ExpType,
+                 properties?: { [prop: string]: ExpType },
+                 methods?: { [name: string]: Method },
+                 params?: string[]): Schema {
     properties = properties || {};
-
+    methods = methods || {};
     derivedFrom = derivedFrom || typeType(Object.keys(properties).length > 0 ? 'Object' : 'Any');
+
     var s: Schema = {
-      derivedFrom: <ExpType | ExpUnionType> derivedFrom,
+      derivedFrom: <ExpType> derivedFrom,
       properties: properties,
-      methods: methods
+      methods: methods,
     };
-    this.register('schema', name, s);
-  }
-
-  isDerivedFrom(descendant: string, ancestor: string, visited?: { [name: string]: boolean}) {
-    if (!visited) {
-      visited = {};
+    // Only add params if this is a Generic schema.
+    if (params) {
+      s.params = params;
     }
-    if (visited[descendant]) {
-      return false;
-    }
-    visited[descendant] = true;
+    return <Schema> this.register('schema', name, s);
+  }
 
-    if (descendant === ancestor) {
+  isDerivedFrom(type: ExpType, ancestor: string): boolean {
+    if (ancestor === 'Any') {
       return true;
     }
 
-    var schema = this.schema[descendant];
-    if (!schema) {
-      return false;
+    switch (type.type) {
```

Comment:
these two *almost* identical switch statements in a row, but one returns in the middle for a certain condition read badly. Can we think of a better way of expressing isDerivedFrom that doesn't have quite so many non-linear control flow kick outs